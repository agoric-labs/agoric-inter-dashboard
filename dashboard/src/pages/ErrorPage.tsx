import { useRouteError } from 'react-router-dom';
import { PageHeader } from '@/components/PageHeader';
import { PageContent } from '@/components/PageContent';
import { MainLayout } from '@/components/MainLayout';

export function ErrorPage() {
  const error: any = useRouteError();

  return (
    <MainLayout>
      <PageHeader title="Oops!" />
      <PageContent>
        <p>Sorry, an unexpected error has occurred.</p>
        <p>
          <i>{error.statusText || error.message}</i>
        </p>
      </PageContent>
    </MainLayout>
  );
}
