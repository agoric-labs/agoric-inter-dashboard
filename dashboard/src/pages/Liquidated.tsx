import { ValueCardGrid } from '@/components/ValueCardGrid';
import { PageHeader } from '@/components/PageHeader';
import { PageContent } from '@/components/PageContent';
import { LiquidatedVaults } from '@/widgets/LiquidatedVaults';
import { VaultStatesChart } from '@/widgets/VaultStatesChart';
import { LiquidatedVaultCountCard } from '@/widgets/LiquidatedVaultCountCard';

export function Liquidated() {
  return (
    <>
      <PageHeader title="Liquidated Vaults" />
      <PageContent>
        <ValueCardGrid>
          <LiquidatedVaultCountCard />
        </ValueCardGrid>
        <VaultStatesChart />
        <hr className="my-5" />
        <LiquidatedVaults />
      </PageContent>
    </>
  );
}
