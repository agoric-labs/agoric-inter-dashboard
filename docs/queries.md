## Check integrity

```sql
with mm as (
   select min(block_height) mn, max(block_height) mx
     from `xxx.raw_cosmos_hub.blocks`
)
select avg(events.count / b.event_count) as status_events
     , avg(msgs.count / b.message_count) as status_msgs
     , avg(txs.count / b.transaction_count) as status_txs
     , count(block_height) / count(distinct block_height) as status_blocks
     , count(block_height) / (max(block_height) - min(block_height)) as status_missing_total
     , sum(events.count) event_count
     , sum(msgs.count) msg_count
     , sum(txs.count) tx_count
     , count(block_height) as total_count
     , (max(block_height) - min(block_height)) as expected_total
     , (max(block_time) - min(block_time)) / count(block_height) avg_build_block_time
     , (max(_sdc_batched_at) - min(_sdc_batched_at)) / count(block_height) avg_extract_block_time
     , max(block_height) as last_block_height
     , max(block_height - prev_block_height) as max_hole
from (
  select block_height
       , block_time
       , event_count
       , message_count
       , transaction_count
       , _sdc_batched_at
       , lag(block_height) over (order by block_height) as prev_block_height
    from `xxx.raw_cosmos_hub.blocks`
    where block_height between (select mn from mm) + 500 and (select mx from mm) - 500
) b
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.events`
   group by 1
) events using (block_height)
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.messages`
   group by 1
) msgs using (block_height)
left join (
  select block_height, count(block_height) count
    from `xxx.raw_cosmos_hub.transactions`
   group by 1
) txs using (block_height)
```

## Fill missing blocks

```
kubectl port-forward agoric-mainnet-tendermint-head-5f876fb589-cvt5c 3333:3333
```

```
with mm as (
   select min(block_height) mn, max(block_height) mx
     from `xxx.raw_cosmos_hub.blocks`
)
select array_to_string(array_agg('curl -d \'' || to_json_string(to_json(struct(earliest, latest))) || '\' http://localhost:3333'), '\n'), count(earliest)
  from (
    select next as earliest
         , least(next + 99, block_height - 1) as latest
     from (
      select block_height
           , lag(block_height) over (order by block_height) as prev_block_height
        from `xxx.raw_cosmos_hub.blocks`
       where block_height between (select mn from mm) + 500 and (select mx from mm) - 500
     )
     cross join unnest(generate_array(prev_block_height + 1, block_height - 1, 100)) next
     where block_height - prev_block_height > 1
     order by 1
  )
```
