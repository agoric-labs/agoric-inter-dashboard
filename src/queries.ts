import { getDateKey, range } from "./utils";

export const PSM_DASHBOARD_QUERY = `
query {
    _metadata {
        lastProcessedHeight
    }
    psmMetrics {
        nodes {
            blockHeight
            denom
            mintedPoolBalance
        }
    }
    psmGovernances {
        nodes {
            mintLimit
            denom
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`;

export const PSM_GRAPH_TOKENS_QUERY = `
query {
    psmMetrics {
        nodes {
            id
            denom
        }
    }
}`;

export const PSM_TOKEN_DAILY_MINT_QUERY = (tokens: Array<string>) => `
query {
    ${tokens?.map((token: string) => `
        ${token}: psmMetricsDailies(first: 90, filter: {denom: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
                id
                denom
                dateKey
                blockTimeLast
                totalMintedProvidedLast
            }
        }`
)}
}`
const currentDate = new Date();
export const VAULTS_DASHBOARD_QUERY = `
query {
    vaultManagerMetrics {
        nodes {
            id
            numActiveVaults
            totalCollateral
            totalDebt
            liquidatingCollateralBrand
        }
    }
    oraclePrices {
        nodes {
            priceFeedName
            typeOutAmount
            typeInAmount
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
    vaultManagerGovernances {
        nodes {
            id
            debtLimit
            liquidationMarginNumerator
            liquidationMarginDenominator
        }
    }
    vaults(filter: {state: {equalTo: "active"}}) {
        nodes {
            id
            denom
            balance
            state
            debt
            blockTime
        }
        totalCount
    }
    oraclePriceDailies(filter: {
        dateKey: { 
            in: [${getDateKey(currentDate).key}, ${getDateKey(currentDate, 1).key}] 
        }
    }) {
        nodes {
            typeInName
            typeInAmountLast
            typeOutAmountLast
            dateKey
        }
    }
}`;

export const OPEN_VAULTS_NEXT_PAGES_QUERY = (pages: number) => `
query {
    ${range(pages).map(index => `v${index}:vaults(filter: {state: {equalTo: "active"}}, first: 100, offset:${(index + 1) * 100}) {
        nodes {
            id
            denom
            balance
            state
            debt
            blockTime
        }
    }`)}
}`

export const VAULTS_DAILY_METRICS_QUERY = (tokens: string[]) => `
query {
    ${tokens?.map((token) =>
    `${token}: vaultManagerMetricsDailies(first: 90, filter:{liquidatingCollateralBrand: {equalTo: "${token}"} }, orderBy:DATE_KEY_DESC)  {
        nodes {
            id
            dateKey
            liquidatingCollateralBrand
            blockTimeLast
            totalCollateralLast
            metricsCount
            totalDebtLast
        }
    }
    ${token}_oracle: oraclePriceDailies (first: 90, filter:{typeInName: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC ) {
        nodes {
            id
            dateKey
            blockTimeLast
            typeInName
            typeInAmountLast
            typeOutAmountLast
        }
    }`)}
}`

export const RESERVE_DASHBOARD_QUERY = `
query {
    reserveMetrics {
        nodes {
            shortfallBalance
            allocations {
                nodes {
                    id
                    denom
                    value
                }
            }
        }
    }
    oraclePrices {
        nodes {
            typeInName
            typeOutAmount
            typeInAmount
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`


export const RESERVE_GRAPH_TOKENS_QUERY = `
query {
    reserveMetrics {
        nodes {
            allocations {
                nodes {
                    id
                    denom
                }
            }
        }
    }
}`

export const RESERVE_DAILY_METRICS_QUERY = (tokens: string[], startDate: number) => `
query {
    ${tokens?.map((token) =>
    `${token}: reserveAllocationMetricsDailies (first: 90, filter: {denom: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
            id
            blockTimeLast
            dateKey
            valueLast
            denom
        }
    }
    ${token}_oracle: oraclePriceDailies (first: 90, filter: {typeInName: {equalTo: "${token}"}}, orderBy:DATE_KEY_DESC) {
        nodes {
          dateKey
          blockTimeLast
          typeInName
          typeInAmountLast
          typeOutAmountLast
        }
    }
    ${token}_last: reserveAllocationMetricsDailies (first: 1, filter: {denom: {equalTo: "${token}"}, valueLast: {greaterThan: "0" }, dateKey: {lessThan: ${startDate}}}, orderBy:DATE_KEY_DESC) {
        nodes {
          id
          blockTimeLast
          dateKey
          valueLast
          denom
        }
    }`)}
}`

export const INTER_DASHBOARD_QUERY = `
query {
    psmMetrics {
        nodes {
            denom
            mintedPoolBalance
            anchorPoolBalance
            mintedPoolBalance
        }
    }
    psmGovernances {
        nodes {
            denom
            mintLimit
        }
    }
    vaultManagerMetrics {
        nodes {
            id
            liquidatingCollateralBrand
            liquidatingDebtBrand
            totalDebt
            totalCollateral
        }
    }
    vaultManagerGovernances {
        nodes {
            id
            debtLimit
        }
    }
    reserveMetrics {
        nodes {
            totalFeeBurned
            shortfallBalance
            allocations {
                nodes {
                    id
                    denom
                    value
                }
            }
        }
    }
    wallets {
        totalCount
    }
    oraclePrices {
        nodes {
            id
            typeInName
            typeOutName
            typeInAmount
            typeOutAmount
            priceFeedName
        }
    }
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
}`

export const LIQUIDATIONS_DASHBOARD = `
query {
    boardAuxes {
        nodes {
            allegedName
            decimalPlaces
        }
    }
    vaultManagerMetrics {
        nodes {
            id
            numLiquidationsCompleted
            liquidatingCollateralValue
            numActiveVaults
            numLiquidationsCompleted
            numLiquidationsAborted
            numLiquidatingVaults
        }
    }
    vaultLiquidations (filter: {state: {equalTo: "liquidated"}}) {
        nodes {
            id
            denom
            debt
            state
            balance
            blockTime
            oraclePrice
            vaultManagerGovernance
            currentState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
            liquidatingState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
        }
    }
}`;

export const VAULT_STATE_DAILIES_QUERY = `
query {
    vaultStatesDailies  ( first: 90, orderBy: ID_DESC, offset: 1) {
        nodes {
            id
            blockTimeLast
            active
            closed
            liquidated
            liquidating
            liquidatedClosed
        }
    }
}`
