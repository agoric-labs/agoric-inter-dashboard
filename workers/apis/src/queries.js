import { range } from './utils';

export const graphqlQuery = {
  query: `
    {
      _metadata{
        lastProcessedHeight
        lastProcessedTimestamp
      }
      oraclePrices {
        nodes {
          typeInName
          typeOutName
          typeInAmount
          typeOutAmount
          blockTime
          blockHeight
        }
      }
      oraclePriceDailies (filter: {dateKey: { greaterThan: 20240501 } }, orderBy: DATE_KEY_DESC) {
        nodes {
          typeInName
          typeOutName
          typeInAmountLast
          typeOutAmountLast
          blockTimeLast
          blockHeightLast
        }
        totalCount
      }
      vaultManagerMetrics {
        nodes {
          id
          blockTime
          blockHeight
          totalCollateral
          totalDebt
          totalCollateralSold
          totalShortfallReceived
          liquidatingCollateralValue
          liquidatingDebtBrand
          liquidatingCollateralBrand
        }
      }
      vaultManagerGovernances  {
        nodes {
          id
          blockTime
          blockHeight
          debtLimit
          liquidationMarginNumerator
          liquidationMarginDenominator
        }
      }
      vaults (filter: {state: {equalTo: "active"}}) {
        nodes {
          id
          blockTime
          blockHeight
          denom
          balance
          state
          debt
          lockedValue
          coin
        }
        totalCount
      }
      vaultLiquidations (filter: {state: {equalTo: "liquidated"}}) {
        nodes {
            id
            blockTime
            blockHeight
            denom
            debt
            state
            balance
            currentState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
            liquidatingState {
                id
                denom
                debt
                state
                balance
                blockTime
            }
        }
        totalCount
    }
    }`,
};

export const PAGINATINATED_DATA_QUERIES = (vaultsPagesCount, vaultLiquidationsPagesCount, oraclePriceDailiesPagesCount) => ({
  query: `{
    ${range(vaultsPagesCount).map(
      (index) => `vaults_${index}:vaults (filter: {state: {equalTo: "active"}}, first: 100, offset:${(index + 1) * 100}) {
      nodes {
        id
        blockTime
        blockHeight
        denom
        balance
        state
        debt
        lockedValue
        coin
      }
    }`
    )}
    ${range(vaultLiquidationsPagesCount).map(
      (index) => `vaultLiquidations_${index}:vaultLiquidations (filter: {state: {equalTo: "liquidated"}}, first: 100, offset:${
        (index + 1) * 100
      }) {
      nodes {
        id
        blockTime
        blockHeight
        denom
        debt
        state
        balance
        currentState {
            id
            denom
            debt
            state
            balance
            blockTime
        }
        liquidatingState {
            id
            denom
            debt
            state
            balance
            blockTime
        }
      }
    }`
    )}    
    ${range(oraclePriceDailiesPagesCount).map(
      (
        index
      ) => `oraclePriceDailies_${index}:oraclePriceDailies (filter:  {dateKey: { greaterThan: 20240501 } }, orderBy: DATE_KEY_DESC, first: 100, offset:${
        (index + 1) * 100
      }) {
      nodes {
        typeInName
        typeOutName
        typeInAmountLast
        typeOutAmountLast
        blockTimeLast
        blockHeightLast
      }
    }`
    )}
  }`,
});

export const LIQUIDATION_ORACLE_PRICES_DAILIES_QUERY = (tokens) => ({
  query: `
{
    ${Object.entries(tokens).map(
      ([token, oracleKeys]) =>
        `${token}:oraclePriceDailies (filter: {
        dateKey: { in: [${oracleKeys.join(', ')}] }, typeInName: {equalTo: "${token}"}
    }) {
        nodes {
            id
            typeInName
            typeOutName
            typeInAmountLast
            typeOutAmountLast
            dateKey
        }
    }`
    )}
}`,
});
