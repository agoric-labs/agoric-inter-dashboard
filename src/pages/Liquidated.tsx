import useSWR from 'swr';
import { AxiosError, AxiosResponse } from 'axios';
import { ValueCardGrid } from '@/components/ValueCardGrid';
import { PageHeader } from '@/components/PageHeader';
import { PageContent } from '@/components/PageContent';
import { LiquidatedVaults } from '@/widgets/LiquidatedVaults';
import { VaultStatesChart } from '@/widgets/VaultStatesChart';
import { LiquidatedVaultCountCard } from '@/widgets/LiquidatedVaultCountCard';
import { subQueryFetcher, subQueryGraphFetcher } from '@/utils';
import { LIQUIDATIONS_DASHBOARD, LIQUIDATION_GRAPH_TOKENS_QUERY, LIQUIDATION_DAILY_METRICS_QUERY } from '@/queries';

export type VaultManagerMetricsNode = {
  id: string;
  liquidatingCollateralValue: number;
  numActiveVaults: number;
  numLiquidatingVaults: number;
  numLiquidationsAborted: number;
  numLiquidationsCompleted: number;
};
type VaultManagerGovernancesNode = {
  id: string;
  liquidationMarginDenominator: number;
  liquidationMarginNumerator: number;
};
type VaultsNode = {
  balance: number;
  debt: number;
  id: string;
  state: string;
  token: string;
};

type LiquidationDashboardResponse = {
  vaults: { nodes: Array<VaultsNode> };
  vaultManagerGovernances: {
    nodes: Array<VaultManagerGovernancesNode>;
  };
  vaultManagerMetrics: {
    nodes: Array<VaultManagerMetricsNode>;
  };
};
export type LiquidationDashboardData = {
  vaults: Array<VaultsNode>;
  vaultManagerGovernances: { [key: string]: VaultManagerGovernancesNode };
};

type TokenNamesResponse = {
  vaultManagerMetrics: {
    nodes: Array<{ id: string; liquidatingCollateralBrand: string }>;
  };
};
type LiquidationMetricsDailyNode = {
  id: string;
  dateKey: number;
  blockTimeLast: string;
  numActiveVaultsLast: number;
  numLiquidatingVaultsLast: number;
  numLiquidationsCompletedLast: number;
  numLiquidationsAbortedLast: number;
  liquidatingCollateralBrand: string;
};

type LiquidationMetricsDailyResponse = {
  [key: string]: {
    nodes: Array<LiquidationMetricsDailyNode>;
  };
};
type GraphData = { key: number; x: string; active: object; liquidated: object };
const sum = (items: Array<string>) => items.reduce((agg, next) => agg + Number(next), 0);

export function Liquidated() {
  const { data, isLoading } = useSWR<AxiosResponse, AxiosError>(LIQUIDATIONS_DASHBOARD, subQueryFetcher);
  const response: LiquidationDashboardResponse = data?.data?.data;

  const vaultManagerGovernances: { [key: string]: VaultManagerGovernancesNode } =
    response?.vaultManagerGovernances?.nodes?.reduce((agg, node) => {
      const idSegments = node?.id?.split('.');
      if (idSegments.length < 4) {
        throw new Error(`Node ID does not contain enough segments: ${node.id}`);
      }
      const managerName = idSegments.slice(0, 4).join('.');
      return { ...agg, [managerName]: node };
    }, {});
  const liquidationDashboardData = {
    vaultManagerGovernances,
    vaults: response?.vaults?.nodes,
  };

  //  Queries for graph
  const { data: tokenNamesData } = useSWR<AxiosResponse, AxiosError>(
    LIQUIDATION_GRAPH_TOKENS_QUERY,
    subQueryGraphFetcher,
  );
  const tokenNamesResponse: TokenNamesResponse = tokenNamesData?.data.data;
  const tokenNames = tokenNamesResponse?.vaultManagerMetrics.nodes.map((node) => node.liquidatingCollateralBrand) || [];

  const { data: dailyMetricsData, isLoading: graphDataIsLoading } = useSWR<AxiosResponse, AxiosError>(
    LIQUIDATION_DAILY_METRICS_QUERY(tokenNames),
    subQueryGraphFetcher,
  );
  const dailyMetricsResponse: LiquidationMetricsDailyResponse = dailyMetricsData?.data.data;
  const graphDataMap: { [key: number]: GraphData } = {};
  Object.values(dailyMetricsResponse || {}).forEach((tokenDataList) => {
    tokenDataList?.nodes.forEach((dailyTokenMetrics) => {
      graphDataMap[dailyTokenMetrics.dateKey] = {
        ...graphDataMap[dailyTokenMetrics.dateKey],
        x: dailyTokenMetrics.blockTimeLast.split('T')[0],
        key: dailyTokenMetrics.dateKey,
        active: {
          ...graphDataMap[dailyTokenMetrics.dateKey]?.active,
          [dailyTokenMetrics.liquidatingCollateralBrand]: dailyTokenMetrics.numActiveVaultsLast,
        },
        liquidated: {
          ...graphDataMap[dailyTokenMetrics.dateKey]?.liquidated,
          [dailyTokenMetrics.liquidatingCollateralBrand]: dailyTokenMetrics.numLiquidationsCompletedLast,
        },
      };
    });
  });
  const sortedGraphDataList = Object.values(graphDataMap);
  sortedGraphDataList.sort((a, b) => a.key - b.key);
  let prevValue: GraphData = sortedGraphDataList[0];
  const graphDataList: Array<GraphData> = sortedGraphDataList.reduce(
    (aggArray: Array<GraphData>, graphData: GraphData) => {
      // filling in missing days
      const prevDay = new Date(prevValue.x);
      const nextDay = new Date(graphData.x);
      const timeDiff = nextDay.getTime() - prevDay.getTime();
      const diffDays = Math.ceil(timeDiff / (1000 * 60 * 60 * 24));
      const missingDays =
        diffDays > 1
          ? Array.from(Array(diffDays - 1).keys()).map((idx) => {
              const newDate = new Date(prevDay);
              newDate.setDate(prevDay.getDate() + 1 + idx);
              const newDateString = newDate.toISOString().slice(0, 10);
              const dateKey = Number(newDateString.replaceAll('-', ''));
              return { ...prevValue, x: newDateString, key: dateKey };
            })
          : [];

      const newAggArray = [...aggArray, ...missingDays, { ...prevValue, ...graphData }];
      prevValue = { ...prevValue, ...graphData };
      return newAggArray;
    },
    [],
  );
  const summedGraphDataList = graphDataList.map((item) => ({
    ...item,
    active: sum(Object.values(item.active)),
    liquidated: sum(Object.values(item.active)),
  }));

  return (
    <>
      <PageHeader title="Liquidated Vaults" />
      <PageContent>
        <ValueCardGrid>
          <LiquidatedVaultCountCard data={response?.vaultManagerMetrics?.nodes} isLoading={isLoading} />
        </ValueCardGrid>
        <VaultStatesChart data={summedGraphDataList} isLoading={graphDataIsLoading} />
        <hr className="my-5" />
        <LiquidatedVaults data={liquidationDashboardData} isLoading={isLoading} />
      </PageContent>
    </>
  );
}
