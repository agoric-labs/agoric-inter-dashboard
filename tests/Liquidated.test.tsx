import React from 'react';
import useSWR from 'swr';
import { render } from '@testing-library/react';
import { Liquidated } from '@/pages/Liquidated';
import { LIQUIDATIONS_DASHBOARD, LIQUIDATION_GRAPH_TOKENS_QUERY } from '@/queries';
import { graphDataMock, dashboardDataMock, graphNodesDataMock } from './__mocks__/Liquidated';

jest.mock('swr', () => ({
  default: jest.fn(),
}));


global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

jest.mock('recharts', () => {
  const OriginalModule = jest.requireActual('recharts');
  return {
    ...OriginalModule,
    ResponsiveContainer: ({ children }: { children: typeof React.Children }) => (
      <OriginalModule.ResponsiveContainer width={800} height={800}>
        {children}
      </OriginalModule.ResponsiveContainer>
    ),
  };
});

describe('Liquidation Dashboard Snapshot tests', () => {
  it('should match snapshot when data is loaded', async () => {
    (useSWR as any).mockImplementation((query: string) => {
      const mockData =
        {
          [LIQUIDATIONS_DASHBOARD]: dashboardDataMock,
          [LIQUIDATION_GRAPH_TOKENS_QUERY]: graphNodesDataMock,
        }[query] || graphDataMock;
      return { data: mockData, error: null, isLoading: false };
    });

    const { container } = render(<Liquidated />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when data is loading', async () => {
    (useSWR as any).mockImplementation(() => ({ data: undefined, error: null, isLoading: true }));

    const { container } = render(<Liquidated />);
    expect(container).toMatchSnapshot();
  });
});
