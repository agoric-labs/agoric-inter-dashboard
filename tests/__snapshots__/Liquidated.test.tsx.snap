// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Liquidation Dashboard Snapshot tests should match snapshot when data is loaded 1`] = `
<div>
  <h1
    class="flex items-center space-x-4 lg:space-x-6 border-b p-8 text-3xl font-bold tracking-tight"
  >
    Liquidated Vaults
  </h1>
  <div
    class="p-8"
  >
    <div
      class="grid grid-cols-2 xl:grid-cols-3 gap-4 mb-5"
    >
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Liquidated Vaults
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            78
          </div>
        </div>
      </div>
    </div>
    <hr
      class="my-5"
    />
    <h2
      class="text-2xl fontebold tracking-tight mb-5"
    >
      Liquidated Vaults
    </h2>
    <div
      class="rounded-md border"
    >
      <div
        class="w-full overflow-auto"
      >
        <table
          class="w-full caption-bottom text-sm"
        >
          <thead
            class="[&_tr]:border-b"
          >
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
            >
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                style="width: 50px;"
              >
                <button
                  class="text-left w-full"
                  type="button"
                >
                  ID
                </button>
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                style="width: 300px;"
              >
                <button
                  class="text-left w-full"
                  type="button"
                >
                  Vault Manager
                </button>
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                style="width: 150px;"
              >
                <button
                  class="text-left w-full"
                  type="button"
                >
                  State
                </button>
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                style="width: 150px;"
              >
                <button
                  class="text-left w-full"
                  type="button"
                >
                  Liquidation Start Time
                </button>
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                style="width: 150px;"
              >
                <button
                  class="text-left w-full"
                  type="button"
                >
                  Liquidation Time
                </button>
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                style="width: 150px;"
              >
                <button
                  class="text-left w-full"
                  type="button"
                >
                  <div
                    class="text-right"
                  >
                    IST Debt Amount
                  </div>
                </button>
              </th>
              <th
                class="h-12 px-4 text-left align-middle font-medium text-muted-foreground [&:has([role=checkbox])]:pr-0"
                style="width: 150px;"
              >
                <button
                  class="text-left w-full"
                  type="button"
                >
                  <div
                    class="text-right"
                  >
                    Collateral Returned Amount
                  </div>
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            class="[&_tr:last-child]:border-0"
          >
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              data-state="false"
            >
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                44
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                ATOM
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                Liquidated
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                2023-08-17 23:00
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                24 minutes
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  25.125
                </div>
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  4.042
                </div>
              </td>
            </tr>
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              data-state="false"
            >
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                83
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                stATOM
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                Liquidated
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                2024-04-13 21:00
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                23 minutes
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  26.565
                </div>
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  3.749
                </div>
              </td>
            </tr>
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              data-state="false"
            >
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                24
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                stkATOM
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                Liquidated
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                2023-11-21 23:00
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                12 hours
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  210.249
                </div>
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  16.157
                </div>
              </td>
            </tr>
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              data-state="false"
            >
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                38
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                stOSMO
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                Liquidated (Closed)
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                2024-04-19 03:00
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                1 hour
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  0.001
                </div>
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  0.001
                </div>
              </td>
            </tr>
            <tr
              class="border-b transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted"
              data-state="false"
            >
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                6
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                stTIA
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                Liquidated
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                2024-04-12 20:00
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                24 minutes
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  7,568.66
                </div>
              </td>
              <td
                class="p-4 align-middle [&:has([role=checkbox])]:pr-0"
              >
                <div
                  class="text-right font-medium"
                >
                  921.651
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Liquidation Dashboard Snapshot tests should match snapshot when data is loading 1`] = `
<div>
  <h1
    class="flex items-center space-x-4 lg:space-x-6 border-b p-8 text-3xl font-bold tracking-tight"
  >
    Liquidated Vaults
  </h1>
  <div
    class="p-8"
  >
    <div
      class="grid grid-cols-2 xl:grid-cols-3 gap-4 mb-5"
    >
      <div
        class="rounded-lg border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="p-6 flex flex-row items-center justify-between space-y-0 pb-2"
        >
          <h3
            class="tracking-tight text-sm font-medium"
          >
            Total Liquidated Vaults
          </h3>
        </div>
        <div
          class="p-6 pt-0"
        >
          <div
            class="text-2xl font-bold"
          >
            <div
              class="animate-pulse bg-muted w-[50px] h-[32px] rounded-full"
            />
          </div>
        </div>
      </div>
    </div>
    <hr
      class="my-5"
    />
    <h2
      class="text-2xl fontebold tracking-tight mb-5"
    >
      Liquidated Vaults
    </h2>
    <div
      class="animate-pulse bg-muted w-full h-[20px] rounded-full mb-2"
    />
    <div
      class="animate-pulse bg-muted w-full h-[20px] rounded-full mb-2"
    />
    <div
      class="animate-pulse bg-muted w-full h-[20px] rounded-full mb-2"
    />
    <div
      class="animate-pulse bg-muted w-full h-[20px] rounded-full mb-2"
    />
  </div>
</div>
`;
